R/cancensus.R

```diff
@@ -366,62 +378,107 @@ list_census_vectors <- function(dataset, use_cache=TRUE) {
 #'
 #' @param searchterm The term to search for e.g. \code{"Ojibway"}.
 #' Search terms are case insensitive. If unable to find a given search term,
 #' this function will suggest the correct spelling to use when possible.
 #' @param dataset The dataset to query for available vectors, e.g.
 #'   \code{"CA16"}.
+#' @param type One of \code{NA}, \code{'Total'}, \code{'Male'} or \code{'Female'}.
+#' If specified, only return variables of specified `type`.
+#'   \code{"CA16"}.
 #'
 #' @export
 #'
 #' @examples
 #' search_census_vectors('Ojibway', 'CA16')
 #'
 #' # This will return a warning that no match was found, but will suggest similar terms.
-#' search_census_vectors('Ojibwe', 'CA16')
-search_census_vectors <- function(searchterm, dataset) {
+#' search_census_vectors('Ojibwe', 'CA16', 'Total')
+search_census_vectors <- function(searchterm, dataset, type=NA) {
   #to do: add caching of vector list here
   veclist <- list_census_vectors(dataset)
-  sublist <- veclist[grep(searchterm, veclist$label, ignore.case = TRUE),]
-  result <- sublist
-  # This part below is extremely inelegant and I look forward to someone adjusting it.
-  # Depth was tested on language for CA16 and CA11, as it looks like language has the most deeply nested variables. 
-  if (any(!is.na(sublist$parent_vector))) {
-    parlist <- veclist[match(sublist$parent_vector, veclist$vector),]
-    result$parent1 <- parlist$label
-    if (any(!is.na(parlist$parent_vector))) {
-      par2list <- veclist[match(parlist$parent_vector, veclist$vector),]
-      result$parent2 <- par2list$label
-      if (any(!is.na(par2list$parent_vector))) {
-        par3list <- veclist[match(par2list$parent_vector, veclist$vector),]
-        result$parent3 <- par3list$label
-        if (any(!is.na(par3list$parent_vector))) {
-          par4list <- veclist[match(par3list$parent_vector, veclist$vector),]
-          result$parent4 <- par4list$label
-          if (any(!is.na(par4list$parent_vector))) {
-            par5list <- veclist[match(par4list$parent_vector, veclist$vector),]
-            result$parent5 <- par5list$label
-            if (any(!is.na(par5list$parent_vector))) {
-              par6list <- veclist[match(par5list$parent_vector, veclist$vector),]
-              result$parent6 <- par6list$label
-              if (any(!is.na(par6list$parent_vector))) {
-                par7list <- veclist[match(par6list$parent_vector, veclist$vector),]
-                result$parent7 <- par7list$label
-              }
-            }
-          }
-        }
-      }
-    }
+  result <- veclist[grep(searchterm, veclist$label, ignore.case = TRUE),]
+
+  # filter by type if needed
+  if (!is.na(type) && length(rownames(result)) > 0) {
+    result <- result[result$type==type,]
   }
+
   # Check if searchterm returned anything
-  if (length(rownames(result)) > 0 ) return(result)
+  if (length(rownames(result)) > 0 ) {
+    attributes(result)$dataset=dataset
```

Comment:
`attr(result, "dataset") <- dataset` is generally preferred to `attributes(result)$dataset <- dataset` (here and elsewhere).